programs/validator-history/src/state.rs

```diff
@@ -524,6 +524,132 @@ impl ValidatorHistory {
+}
+
+impl Default for ClusterHistoryEntry {
+    fn default() -> Self {
+        Self {
+            total_blocks: u32::MAX,
+            epoch: u16::MAX,
+            padding: [u8::MAX; 122],
+        }
+    }
+}
+
+#[zero_copy]
+pub struct CircBufCluster {
+    pub idx: u64,
+    pub is_empty: u8,
+    pub padding: [u8; 7],
+    pub arr: [ClusterHistoryEntry; MAX_ITEMS],
+}
+
+impl CircBufCluster {
+    pub fn push(&mut self, item: ClusterHistoryEntry) {
+        self.idx = (self.idx + 1) % self.arr.len() as u64;
+        self.arr[self.idx as usize] = item;
+        self.is_empty = 0;
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.is_empty == 1
+    }
+
+    pub fn last(&self) -> Option<&ClusterHistoryEntry> {
+        if self.is_empty() {
+            None
+        } else {
+            Some(&self.arr[self.idx as usize])
+        }
+    }
+
+    pub fn last_mut(&mut self) -> Option<&mut ClusterHistoryEntry> {
+        if self.is_empty() {
+            None
+        } else {
+            Some(&mut self.arr[self.idx as usize])
+        }
+    }
+
+    pub fn arr_mut(&mut self) -> &mut [ClusterHistoryEntry] {
+        &mut self.arr
+    }
+
+    /// Returns &ClusterHistoryEntry for each existing entry in range [start_epoch, end_epoch], factoring for wraparound
+    /// Returns None if either start_epoch or end_epoch is not in the CircBuf
+    pub fn epoch_range(&self, start_epoch: u16, end_epoch: u16) -> Vec<&ClusterHistoryEntry> {
+        // creates an iterator that lays out the entries in consecutive order, handling wraparound
+        self.arr[(self.idx as usize + 1)..]
+            .iter()
+            .chain(self.arr[..=(self.idx as usize)].iter())
+            .filter(|entry| entry.epoch >= start_epoch && entry.epoch <= end_epoch)
+            .collect()
+    }
+
+    pub fn total_blocks_latest(&self) -> Option<u32> {
+        if let Some(entry) = self.last() {
+            if entry.total_blocks != ClusterHistoryEntry::default().total_blocks {
+                return Some(entry.total_blocks);
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn total_blocks_range(&self, start_epoch: u16, end_epoch: u16) -> Vec<Option<u32>> {
+        let epoch_range = self.epoch_range(start_epoch, end_epoch);
+        epoch_range
+            .iter()
+            .map(|entry| {
+                if entry.total_blocks != ClusterHistoryEntry::default().total_blocks {
+                    Some(entry.total_blocks)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<Option<u32>>>()
+    }
+}
+
+impl ClusterHistory {
+    pub const SIZE: usize = 8 + size_of::<Self>();
+    pub const MAX_ITEMS: usize = MAX_ITEMS;
+    pub const SEED: &'static [u8] = b"cluster-history";
+
+    // Sets total blocks for the target epoch
+    pub fn set_blocks(&mut self, epoch: u16, blocks_in_epoch: u32) -> Result<()> {
```

Comment:
this is going to mess up on backfill

Reply:
added check to ensure updates are sequential in backfill ix